### PR TITLE
Hacktoberfest: Update Linked List - `print_reverse` method

### PR DESCRIPTION
**Pull request by shermanhui**
### **Describe your change:**

Update `make_linked_list` helper to use a generator expression instead of 
slicing the original `elements_list` to avoid creating a temporary list
to provide an example that works in O(1) space and O(n) time complexity.
This change is on [L46](https://github.com/TheAlgorithms/Python/pull/2792/files#diff-36ddde92b9f9d5f465b51b42a0ff4109R46).

I've also added doctests to this file to add test coverage for 
expected behaviour and edge cases. When adding the tests, I also
added an additional type check in `print_reverse` [L58](https://github.com/TheAlgorithms/Python/pull/2792/files#diff-36ddde92b9f9d5f465b51b42a0ff4109R58) as I discovered
the function would break when providing an argument that is not a
`Node` instance.

I've checked off "Fix a bug or typo in an existing algorithm" as it 
seems to be the option that closest reflects what I've done in this PR.

* [ ] Add an algorithm?
* [x] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [x] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [ ] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by shellhub** [comment]
@shermanhui I formatted the code. Hope you do not mind

**Review by shellhub — APPROVED** [review]

File: data_structures/linked_list/print_reverse.py [hunk 1]
```diff
@@ -8,48 +8,63 @@ def __init__(self, data=None):
...
         raise Exception("The Elements List is empty")
 
-    # Set first element as Head
-    head = Node(elements_list[0])
-    current = head
-    # Loop through elements from position 1
-    for data in elements_list[1:]:
-        current.next = Node(data)
+    current = head = Node(elements_list[0])
+    for i in range(1, len(elements_list)):
```
Comment (shermanhui) [inline data_structures/linked_list/print_reverse.py:37]: @shellhub Thanks for the review and merge, I was hoping to improve the algorithm by using the generator as a part of my Hacktoberfest contribution for #2510. What was the reason for removing it for the regular iteration over a list?

i.e. (what I had before)

```
for data in (elements_list[i] for i in range(1, list_length)):
```

Comment (shellhub) [reply data_structures/linked_list/print_reverse.py:37]: This way need to generate a list and then iterate. It's slow.

Comment (shellhub) [reply data_structures/linked_list/print_reverse.py:37]:
``` python
def fun1() -> None:
    my_list = []
    elements_list = [14, 52, 14, 12, 43]
    for i in range(0, len(elements_list)):
        my_list.append(i)


def fun2() -> None:
    my_list = []
    elements_list = [14, 52, 14, 12, 43]
    for data in (elements_list[i] for i in range(1, len(elements_list))):
        my_list.append(data)


if __name__ == '__main__':
    import timeit

    print(timeit.timeit("fun1()", setup="from __main__ import fun1", number=100000))
    print(timeit.timeit("fun2()", setup="from __main__ import fun2", number=100000))
```
output:
0.07830935700000001
0.10762060600000001

Comment (shermanhui) [reply data_structures/linked_list/print_reverse.py:37]: Thanks, @shellhub that was informative! I only ran `timeit` against the list slicing. I should have checked the regular iteration as well. I guess it's also better to keep it simple! 😃

Comment (anambibi) [reply data_structures/linked_list/print_reverse.py:37]: How we Iris dataset from an online source. Read the dataset into a data structure, preferably
in a 2D array??? Anyone help me please its my assignment

**Review by jessica02200 — APPROVED** [review]

**Pull request merged**